From eb895753d89db0dcda676305843a10d9cf3ce0c0 Mon Sep 17 00:00:00 2001
From: Joerg Hofrichter <joerg.hofrichter@ni.com>
Date: Wed, 29 Jan 2020 10:48:09 +0100
Subject: [PATCH 15/37] board: ni: choose correct ec image based on board rev
 from EEPROM

- ec_image is now automatically set to the correct file based on
  the new variables ec_mboard and ec_compat
- These variables are automatically set during board init based
  on the INFO tag in the MB EEPROM (pid, compat_rev)
- Default values are used if the MB EEPROM is uninitialized
- The default values are set for the right board revision
  depending on the defconfig that was used. For enabling this,
  separate sys config files ni-x410-rev2.h and ni-x410-rev3.h
  are used.

Signed-off-by: Joerg Hofrichter <joerg.hofrichter@ni.com>
---
 board/ni/zynqmp/board.c        | 53 ++++++++++++++++++++++++++++++++++
 configs/ni_x410_rev2_defconfig |  2 +-
 configs/ni_x410_rev3_defconfig |  2 +-
 include/configs/ni-x410-rev2.h | 20 +++++++++++++
 include/configs/ni-x410-rev3.h | 20 +++++++++++++
 include/configs/ni-x410.h      | 14 ++++++++------
 6 files changed, 103 insertions(+), 8 deletions(-)
 create mode 100644 include/configs/ni-x410-rev2.h
 create mode 100644 include/configs/ni-x410-rev3.h

diff --git a/board/ni/zynqmp/board.c b/board/ni/zynqmp/board.c
index aafa8ce24f0..7f0ec6042cb 100644
--- a/board/ni/zynqmp/board.c
+++ b/board/ni/zynqmp/board.c
@@ -206,6 +206,28 @@ unsigned long do_go_exec(ulong (*entry)(int, char * const []), int argc,
 	return ret;
 }
 
+struct xboard_match {
+	u16 pid;
+	const char *name;
+	const char *dtb;
+};
+
+static const struct xboard_match mb_match[] = {
+	{ .pid = 0x0410, .name = "X410 (Titanium)", .dtb = "x410" },
+	{},
+};
+
+static const char *get_mboard_dtb(u16 pid)
+{
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(mb_match); i++)
+		if (pid == mb_match[i].pid)
+			return mb_match[i].dtb;
+
+	return "unsupported";
+}
+
 /*
  * dtbo is a format string containing %s (slot) and %u (rev) in that order
  */
@@ -267,6 +289,36 @@ static void append_overlay(char *overlays, struct eeprom_ctx *ctx, const char *s
 	sprintf(overlays, "%s#conf@%s", overlays, overlay);
 }
 
+static void probe_mboard(struct eeprom_ctx *ctx)
+{
+	const struct usrp_eeprom_board_info *info;
+
+	char variant_buf[128];
+	char compat_rev_buf[8];
+
+	if (!ctx->valid) {
+		printf("Error: MB EEPROM is uninitialized, using default values from environment");
+		return;
+	}
+
+	info = tlv_lookup(ctx->eeprom.tlv, ctx->eeprom.size,
+			  USRP_EEPROM_BOARD_INFO_TAG);
+	if (!info) {
+		printf("Error: MB EEPROM is missing the INFO tag, using default values from environment");
+		return;
+	}
+
+	snprintf(variant_buf, sizeof(variant_buf), "%s-rev%d", get_mboard_dtb(info->pid), info->compat_rev);
+	snprintf(compat_rev_buf, sizeof(compat_rev_buf), "%d", info->compat_rev);
+
+	env_set("variant", variant_buf);
+
+	// TODO: align dtb name to other usrps
+	// this should be: env_set("ec_mboard", get_mboard_dtb(info->pid));
+	env_set("ec_mboard", "titanium");
+	env_set("ec_compat", compat_rev_buf);
+}
+
 int board_late_init(void)
 {
 	char overlays[256] = "";
@@ -308,6 +360,7 @@ int board_late_init(void)
 	load_eeprom("ni,dioaux-eeprom", &dioaux_eeprom);
 	load_eeprom("ni,clkaux-eeprom", &clkaux_eeprom);
 
+	probe_mboard(&mb_eeprom);
 	append_overlay(overlays, &db0_eeprom, "db0");
 	append_overlay(overlays, &db1_eeprom, "db1");
 	append_overlay(overlays, &dioaux_eeprom, "dioaux");
diff --git a/configs/ni_x410_rev2_defconfig b/configs/ni_x410_rev2_defconfig
index 9d53e2fa305..224c30441d2 100644
--- a/configs/ni_x410_rev2_defconfig
+++ b/configs/ni_x410_rev2_defconfig
@@ -1,6 +1,6 @@
 CONFIG_ARM=y
 CONFIG_SYS_VENDOR="ni"
-CONFIG_SYS_CONFIG_NAME="ni-x410"
+CONFIG_SYS_CONFIG_NAME="ni-x410-rev2"
 CONFIG_ARCH_ZYNQMP=y
 CONFIG_SYS_TEXT_BASE=0x8000000
 CONFIG_SYS_MALLOC_F_LEN=0x8000
diff --git a/configs/ni_x410_rev3_defconfig b/configs/ni_x410_rev3_defconfig
index ee7f873ad45..b818c3c38b0 100644
--- a/configs/ni_x410_rev3_defconfig
+++ b/configs/ni_x410_rev3_defconfig
@@ -1,6 +1,6 @@
 CONFIG_ARM=y
 CONFIG_SYS_VENDOR="ni"
-CONFIG_SYS_CONFIG_NAME="ni-x410"
+CONFIG_SYS_CONFIG_NAME="ni-x410-rev3"
 CONFIG_ARCH_ZYNQMP=y
 CONFIG_SYS_TEXT_BASE=0x8000000
 CONFIG_SYS_MALLOC_F_LEN=0x8000
diff --git a/include/configs/ni-x410-rev2.h b/include/configs/ni-x410-rev2.h
new file mode 100644
index 00000000000..40a2c7b3896
--- /dev/null
+++ b/include/configs/ni-x410-rev2.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Configuration for Ettus X410
+ *
+ * (C) Copyright 2019 National Instruments
+ * Michael Auchter <michael.auchter@ni.com>
+ * Joerg Hofrichter <joerg.hofrichter@ni.com>
+ */
+
+#ifndef __CONFIG_ETTUS_X410_REV2_H
+#define __CONFIG_ETTUS_X410_REV2_H
+
+#define EXTRA_ENV_MACHINE_DEFAULTS \
+  "variant=x410-rev2\0" \
+  "ec_mboard=titanium\0" \
+  "ec_compat=2\0"
+
+#include "ni-x410.h"
+
+#endif /* __CONFIG_ETTUS_X410_REV2_H */
diff --git a/include/configs/ni-x410-rev3.h b/include/configs/ni-x410-rev3.h
new file mode 100644
index 00000000000..8cb641cebc5
--- /dev/null
+++ b/include/configs/ni-x410-rev3.h
@@ -0,0 +1,20 @@
+/* SPDX-License-Identifier: GPL-2.0+ */
+/*
+ * Configuration for Ettus X410
+ *
+ * (C) Copyright 2019 National Instruments
+ * Michael Auchter <michael.auchter@ni.com>
+ * Joerg Hofrichter <joerg.hofrichter@ni.com>
+ */
+
+#ifndef __CONFIG_ETTUS_X410_REV3_H
+#define __CONFIG_ETTUS_X410_REV3_H
+
+#define EXTRA_ENV_MACHINE_DEFAULTS \
+  "variant=x410-rev3\0" \
+  "ec_mboard=titanium\0" \
+  "ec_compat=3\0"
+
+#include "ni-x410.h"
+
+#endif /* __CONFIG_ETTUS_X410_REV3_H */
diff --git a/include/configs/ni-x410.h b/include/configs/ni-x410.h
index d74342809cb..2aed4030527 100644
--- a/include/configs/ni-x410.h
+++ b/include/configs/ni-x410.h
@@ -10,13 +10,15 @@
 #ifndef __CONFIG_ETTUS_X410_H
 #define __CONFIG_ETTUS_X410_H
 
+#ifndef EXTRA_ENV_MACHINE_DEFAULTS
+#error undefined
+#endif
+
 #define CONFIG_EXTRA_ENV_SETTINGS \
-        "ec_compat=2\0" \
+        "fit_image=boot/fitImage\0" \
         "ec_disable_swsync=0\0" \
-        "ec_image=lib/firmware/ni/ec-titanium-rev2.RW.bin\0" \
-        "expand_ec_image=setenv ec_image lib/firmware/ni/ec-titanium-rev${ec_compat}.RW.bin\0" \
+        "expand_ec_image=setenv ec_image lib/firmware/ni/ec-${ec_mboard}-rev${ec_compat}.RW.bin\0" \
         "load_addr=0x2000000\0" \
-        "fit_image=boot/fitImage\0" \
         "bootargs=uio_pdrv_genirq.of_id=usrp-uio at24.io_limit=64\0" \
         "emmcboot=" \
           "setenv disk_name eMMC && " \
@@ -32,7 +34,7 @@
           "echo Copying FIT from ${disk_name} to RAM... && " \
           "setenv bootargs root=${kernel_root} ${bootargs} && " \
           "load ${uboot_root} $load_addr ${fit_image} && " \
-          "bootm ${load_addr}#conf@${variant}.dtb${overlays}\0" \
+          "bootm ${load_addr}#conf@ni_${variant}.dtb${overlays}\0" \
         "swsync_disk=if test ${ec_disable_swsync} = 0; then " \
           "echo Copying MCU FW from ${disk_name} to RAM... && " \
           "run expand_ec_image; " \
@@ -52,5 +54,5 @@
           "crosec swsync ${load_addr} ${filesize}; fi;\0" \
         "distro_bootcmd=run $modeboot\0" \
-        "variant=" CONFIG_DEFAULT_DEVICE_TREE "\0"
+        EXTRA_ENV_MACHINE_DEFAULTS
 
 #include <configs/xilinx_zynqmp.h>
 
-- 
2.43.0

